### PR TITLE
Feature/51 bookmarks widget

### LINKED ISSUE #51: Add Bookmarks Widget
[Bookmarks](https://developers.arcgis.com/javascript/latest/api-reference/esri-widgets-Bookmarks.html#properties-summary)

### PR DESCRIPTION
**Pull request by AndersenBell**
closes #51 

Adds support for the bookmark widget.
Adds missing properties on the Expander widget
Adds deprecated comments to deprecated properties
Adds Widget re-render support

### CODE REVIEW

File: src/dymaptic.GeoBlazor.Core/Components/Widgets/ExpandWidget.cs [hunk 1]
```diff
@@ -25,13 +26,15 @@ public class ExpandWidget : Widget
 
     /// <summary>
     ///     Icon font used to style the Expand button when collapsed.
+    ///     Deprecated
     /// </summary>
     [Parameter]
     [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
     public string? CollapseIconClass { get; set; }
```
Comment (seahro) [inline src/dymaptic.GeoBlazor.Core/Components/Widgets/ExpandWidget.cs:33]: Do we want to simply remove these since they are deprecated?

**Review by seahro — APPROVED** [review]
Looks good.  Good job integrating the variety of items that go into implementing each bookmark.  
One question about keeping the Deprecated properties.

File: samples/dymaptic.GeoBlazor.Core.Sample.Shared/Pages/Bookmarks.razor [hunk 2]
```diff
@@ -0,0 +1,53 @@
+@page "/Bookmarks"
+@using dymaptic.GeoBlazor.Core.Components.Widgets;
+
+<PageTitle>Bookmarks and Bookmarks Widget</PageTitle>
+<h1>Bookmarks Widget</h1>
+
+<div class="links-div">
+    <a class="btn btn-secondary" target="_blank" href="https://developers.arcgis.com/javascript/latest/api-reference/esri-widgets-Bookmarks.html">ArcGIS API for JavaScript Reference</a>
+    <a class="btn btn-primary" target="_blank" href="https://www.arcgis.com/home/item.html?id=70b726074af04a7e9839d8a07f64c039">Bookmarks widget</a>
```
Comment (TimPurdum) [inline samples/dymaptic.GeoBlazor.Core.Sample.Shared/Pages/Bookmarks.razor:9]: The text on this line should say "Hurricanes", i.e., the name of the layer.

File: src/dymaptic.GeoBlazor.Core/Components/Widgets/ExpandWidget.cs (hunk 1 above)
Comment (TimPurdum) [reply src/dymaptic.GeoBlazor.Core/Components/Widgets/ExpandWidget.cs:33]: No, @AndersenBell we should use the C# `[Obsolete]` attribute for deprecated properties.

File: samples/dymaptic.GeoBlazor.Core.Sample.Shared/Pages/Bookmarks.razor [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    </Map>
+    @if (_showBookmarks)
+    {
+        <ExpandWidget ContainerId="BookmarkDiv" Position="OverlayPosition.TopRight">
+            <div class="container">
+                <div class="row">
+                    <div class="col-12">
+                        <p>Click on a bookmark to navigate to its location.</p>
+                        <BookmarksWidget  />
+                    </div>
```
Comment (TimPurdum) [inline samples/dymaptic.GeoBlazor.Core.Sample.Shared/Pages/Bookmarks.razor:34]: Is this page not finished? Or does the BookmarksWidget really work with no parameters?

File: src/dymaptic.GeoBlazor.Core/Components/Views/Viewpoint.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+            case Geometry geometry:
+                TargetGeometry = null;
+                break;
+            default:
+                await base.UnregisterChildComponent(child);
+                break;
+        }
+
+    }
+
```
[the remark this answers is not in the document]
Comment (TimPurdum) [reply src/dymaptic.GeoBlazor.Core/Components/Views/Viewpoint.cs:61]: In this case, I don't think any of your children have required props, but it's still good if we do this everywhere.

File: src/dymaptic.GeoBlazor.Core/Components/Widgets/ExpandWidget.cs [hunk 5]
```diff
@@ -110,4 +140,12 @@ public override async Task UnregisterChildComponent(MapComponent child)
                 break;
         }
     }
+}
+
+[JsonConverter(typeof(EnumToKebabCaseStringConverter<Mode>))]
+public enum Mode
```
Comment (TimPurdum) [inline src/dymaptic.GeoBlazor.Core/Components/Widgets/ExpandWidget.cs:146]: this will need an xml doc comment.

**Review by TimPurdum — APPROVED** [review]
A few odds and ends, clarifying the two samples, adding [Obsolete] to deprecated properties, etc.

File: src/dymaptic.GeoBlazor.Core/Scripts/jsBuilder.ts [hunk 6]
```diff
@@ -439,6 +444,35 @@ export function buildJsGeometry(geometry: DotNetGeometry): Geometry | null {
...
 
+export function buildJsBookmark(dnBookmark: DotNetBookmark): Bookmark | null {
+    if (dnBookmark === undefined || dnBookmark === null) return null;
+    let bookmark = new Bookmark();
+    bookmark.name = dnBookmark.name ?? undefined;
+    bookmark.timeExtent = dnBookmark.timeExtent ?? undefined;
+
+    if (!(dnBookmark.thumbnail == null)) {
+        //ESRI has this as an "object" with url property
+        let thumbnail = new Object();
```
[the remark this answers is not in the document]
Comment (AndersenBell) [reply src/dymaptic.GeoBlazor.Core/Scripts/jsBuilder.ts]: thanks, I tried that it couldn't find the syntax to get it to work, works now.

File: src/dymaptic.GeoBlazor.Core/Scripts/jsBuilder.ts [hunk 7]
```diff
@@ -585,7 +622,7 @@ export async function buildJsPopup(dotNetPopup: any, viewId: string): Promise<Po
         autoCloseEnabled: dotNetPopup.autoCloseEnabled ?? false,
         autoOpenEnabled: dotNetPopup.autoOpenEnabled ?? true,
         collapseEnabled: dotNetPopup.collapseEnabled ?? true,
-        collapsed: dotNetPopup.collapsed ?? false,
+        //collapsed: dotNetPopup.collapsed ?? false,
```
Comment (TimPurdum) [inline src/dymaptic.GeoBlazor.Core/Scripts/jsBuilder.ts:625]: hmm, this should probably be reverted? If it's causing problems, we can try `dotNetPopup.collapsed ?? undefined`.

**Review by TimPurdum — APPROVED** [review]
Just a few odds and ends to wrap up, looking good!

File: samples/dymaptic.GeoBlazor.Core.Sample.Shared/Pages/Bookmarks.razor [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        padding: 0;
+        margin: 0;
+        height: 100%;
+        width: 100%;
+    }
+
+    #timeSliderDiv {
+        width: 300px;
+    }
+</style>
```
[the remark this answers is not in the document]
Comment (AndersenBell) [reply samples/dymaptic.GeoBlazor.Core.Sample.Shared/Pages/Bookmarks.razor:20]: this is copied from the ESRI example... so maybe?

File: samples/dymaptic.GeoBlazor.Core.Sample.Shared/Pages/Bookmarks.razor [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+        var layer = MapView!.Map!.Layers.FirstOrDefault();
+        if (_layerView is null) return;
+
+        var hurricaneName = eventArgs.Bookmark.Name;
+
+        var featureEffect = new FeatureEffect
+            {
+                Filter = new FeatureFilter()
+                {
+                    Where = "Name = '" + hurricaneName.ToUpper() + "'"
```
[the remark this answers is not in the document]
Comment (AndersenBell) [reply samples/dymaptic.GeoBlazor.Core.Sample.Shared/Pages/Bookmarks.razor:68]: this is what ESRI did so, I copied the JS, it should be fixed

File: src/dymaptic.GeoBlazor.Core/Components/Layers/FeatureLayerView.cs [hunk 10]
```diff
@@ -48,6 +54,12 @@ public async Task SetFilter(FeatureFilter? filter)
         Filter = filter;
     }
 
+    public async Task SetFeatureEffect(FeatureEffect? featureEffect)
+    {
+        await JsObjectReference!.InvokeVoidAsync("setFeatureEffect", CancellationTokenSource.Token, featureEffect);
+        FeatureEffect = featureEffect;
+    }
```
[the remark this answers is not in the document]
Comment (AndersenBell) [reply src/dymaptic.GeoBlazor.Core/Components/Layers/FeatureLayerView.cs:61]: We can leave that for when we need to use it

File: samples/dymaptic.GeoBlazor.Core.Sample.Shared/Pages/Bookmarks.razor (hunk 8 above)
[the remark this answers is not in the document]
Comment (TimPurdum) [reply samples/dymaptic.GeoBlazor.Core.Sample.Shared/Pages/Bookmarks.razor:20]: I don't think you need the <link> line or the <style> tag.

**Review by seahro — APPROVED** [review]
looks great.  "Effects", "Time" and "Filter" will be nice additions to built out in future iterations.

File: src/dymaptic.GeoBlazor.Core/Components/Widgets/BookmarksWidget.cs [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+    /// A collection of Bookmarks.
+    /// </summary>
+    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
+    public List<Bookmark>? Bookmarks { get; set; } 
+
+    /// <summary>
+    ///     Handler delegate for click events on the view.
+    /// </summary>
+    [Parameter]
+    public EventCallback<BookmarkSelectEvent> OnClick { get; set; }
```
[the remark this answers is not in the document]
Comment (AndersenBell) [reply src/dymaptic.GeoBlazor.Core/Components/Widgets/BookmarksWidget.cs:53]: ok, xml comments and names should be updated

File: src/dymaptic.GeoBlazor.Core/Scripts/jsBuilder.ts (hunk 7 above)
Comment (AndersenBell) [reply src/dymaptic.GeoBlazor.Core/Scripts/jsBuilder.ts:625]: this looks like an artifact from a bad merge or something
image

**Pull request merged**